src/graph/graphcontext.c

```diff
@@ -59,6 +69,18 @@ GraphContext *GraphContext_New(const char *graph_name, size_t node_cap, size_t e
 	gc->encoding_context = GraphEncodeContext_New();
 	gc->decoding_context = GraphDecodeContext_New();
 
+
+	/* Build the cache pool. The cache pool contains a cache for each thread in the thread pool, to avoid congestion.
+	 * Each thread is getting its cache by its thread id. */
```

Comment:
I think it is unwise to maintain 1 cache per thread, and avoiding locks is an inadequate justification for this setup.

As a user I would be very confused to find my repeated queries having seemingly arbitrary cache misses, and after saturating all threads, we'll have N-1 redundant copies of an execution plan.

As an alternative, I think you should consider writing to the cache in a separate worker thread to avoid lock contention on insertion. For cache retrieval, either accept the minimal wait of lock acquisition or, if you really think that is unacceptable, consider solutions like `pthread_mutex_timedlock` or `pthread_mutex_trylock` that give up on cache retrieval.